test/router-luatest/reload_test.lua

```diff
@@ -0,0 +1,263 @@
+local t = require('luatest')
+local vtest = require('test.luatest_helpers.vtest')
+local git_util = require('test.lua_libs.git_util')
+local os = require('os')
+
+local g = t.group('reload_router')
+local cfg_template = {
+    sharding = {
+        {
+            replicas = {
+                replica_1_a = {
+                    master = true,
+                },
+                replica_1_b = {},
+            },
+        },
+        {
+            replicas = {
+                replica_2_a = {
+                    master = true,
+                },
+            },
+        },
+    },
+    bucket_count = 100
+}
+
+local global_cfg
+local latest_hash
+local vshard_copy_path
+
+g.before_all(function()
+    global_cfg = vtest.config_new(cfg_template)
+
+    -- The test works in the following directory
+    vshard_copy_path = vtest.VARDIR..'/vshard_copy_router_luatest'
+    os.execute('mkdir ' .. vshard_copy_path)
+    -- Copy sources to the new location
+    os.execute('cp -rf ' .. vtest.SOURCEDIR .. '/.git ' .. vshard_copy_path)
+
+    -- Latest git log hash of router/init.lua changes.
+    -- Used to test router functionality on the latest router's version.
+    latest_hash = git_util.log_hashes({args = '-1', dir = vtest.SOURCEDIR})[1]
+
+    -- No need to update storages. Just run them on the latest version.
+    vtest.cluster_new(g, global_cfg)
+    vtest.cluster_bootstrap(g, global_cfg)
+    vtest.cluster_rebalancer_disable(g)
+
+    -- Basic storage configuration
+    vtest.cluster_exec_each_master(g, function()
+        box.schema.space.create('test'):create_index('primary')
+    end)
+
+    vtest.cluster_exec_each(g, function()
+        rawset(_G, 'insert', function(space_name, tuple)
+            return box.space[space_name]:insert(tuple)
+        end)
+        rawset(_G, 'get', function(space_name, key)
+            return box.space[space_name]:get(key)
+        end)
+        rawset(_G, 'info', function()
+            local info = box.info
+            return {info.cluster.uuid, info.uuid}
+        end)
+    end)
+end)
+
+g.after_all(function()
+    g.cluster:drop()
+end)
+
+--
+-- Reload test template:
+--     1. Invoke create_router_at:
+--         * Checkout to old version;
+--         * Create and start a router;
+--     3. Test smth on the old version;
+--     4. Invoke reload_router():
+--         * Checkout to the latest version;
+--         * Reload the module.
+--     5. Test smth on the new version
+--     6. Drop a router with vtest.drop_instance
+--
+
+local function create_router_at(hash)
+    git_util.exec('checkout', {args = hash .. ' -f', dir = vshard_copy_path})
+    local router = vtest.router_new(g, 'router-'..hash)
+    router:exec(function(vshard_path)
+        -- Force 'require' to use new directory
+        package.path = string.format('%s/?.lua;%s/?/init.lua;',
+                                     vshard_path, vshard_path)
+        package.loaded['vshard.router'] = nil
```

Comment:
Probably you need to clear all packages started with vshard. Otherwise old vshard.router could use new vshard.consts/utils/...

Reply:
Sounds reasonable. Fixed

Reply:
Seems here you need to do something like:
```
for package_name in pairs(package.loaded) do
    if package_name:startswith('vshard') then
        package.loaded[package_name] = nil
    end
end
```

I don't see such changes

Reply:
I applied the code, you proposed. But I thought  `package.loaded["vshard"] = nil` does the same thing, but with reloading `vshard/init.lua` too. Is it true?